File: implementations/typescript/packages/llmproxy/src/loop.ts
// SPDX-License-Identifier: Apache-2.0
import { canonicalJson, record, type Envelope } from "@psp-cdl/core";
import { verifyEnvelope, type VerificationPolicy } from "@psp-cdl/core/crypto";
import { aggregateCapabilities, evaluateBatch, type PolicyInput } from "@psp-cdl/cdl";
import { SecurityService, ServiceError, identifier, type Principal } from "@psp-cdl/api-server";
import { WorkflowStore, OwnerCoordinator, StoreError, bounded, integer,type OwnerReservation } from "@psp-cdl/api-server/persistence";
import { McpDispatchGate, DispatchError, bindingDigest, type CapabilitySource } from "@psp-cdl/mcpproxy";

export const LLM_LOOP_PROFILE="PSP-LLM-LOOP-0.1";
export class LoopError extends Error {
  constructor(public readonly code:string) {super(code);this.name="LoopError";}
}
export interface LoopOptions {deadline:number;cancelled:()=>boolean;maxSteps:number}
export interface ProviderRegistration {
  id:string;revision:string;sources:CapabilitySource[];complete:boolean;
  invoke(request:Record<string,unknown>,options:Pick<LoopOptions,"deadline"|"cancelled">):unknown|Promise<unknown>;
}
export interface LoopAuthority {
  revision:string;policyVersion:string;providerId:string;providerRevision:string;registryRevision:string;
  expires:number;releaseSources:CapabilitySource[];releaseComplete:boolean;
}
export type LoopPhase="inference"|"tool"|"release";
export interface LoopHost {
  authenticate(token:string):Principal|null|Promise<Principal|null>;
  now():number;
  snapshot(principal:Principal,session:Record<string,unknown>):LoopAuthority|Promise<LoopAuthority>;
  prompt(principal:Principal,binding:Record<string,unknown>):Envelope|Promise<Envelope>;
  verification(principal:Principal,binding:Record<string,unknown>):
    Pick<VerificationPolicy,"keys"|"clockSkew">|Promise<Pick<VerificationPolicy,"keys"|"clockSkew">>;
  policy(principal:Principal,binding:Record<string,unknown>,data:Record<string,unknown>,phase:LoopPhase):
    {bindingDigest:string;resources:PolicyInput[]}|Promise<{bindingDigest:string;resources:PolicyInput[]}>;
  authorizeFinal(principal:Principal,binding:Record<string,unknown>,data:Record<string,unknown>):boolean|Promise<boolean>;
}
const fail=(code:string):never=>{throw new LoopError(code);};
function copy(value:unknown,code="INVALID_REQUEST"):any {try{return bounded(value);}catch{return fail(code);}}
async function callback<T>(fn:()=>T|Promise<T>):Promise<T> {try{return await fn();}catch{return fail("HOST_ERROR");}}
function caps(sources:CapabilitySource[],complete:boolean):string[] {
  try{return aggregateCapabilities(sources,complete);}catch{return fail("INVALID_CAPABILITIES");}
}
/** Required signed attributes. The binding is host-only; do not put it in model messages. */
export function promptContext(binding:Record<string,unknown>):Record<string,string> {
  const fields={"tenant-id":"tenantId","subject-id":"subjectId","session-id":"sessionId","session-version":"sessionVersion",
    "node-id":"nodeId","node-version":"nodeVersion","store-epoch":"epoch","policy-version":"policyVersion",
    "authority-revision":"authorityRevision","provider-id":"providerId","provider-revision":"providerRevision","registry-revision":"registryRevision"};
  return Object.fromEntries(Object.entries(fields).map(([attribute,field])=>[attribute,String(binding[field])]));
}

/** Bounded, buffered, host-embedded loop. No listener, provider SDK or implicit credential lookup. */
export class BufferedLlmLoop {
  private readonly auth:SecurityService;
  protected readonly coordinator:OwnerCoordinator;
  protected readonly provider:ProviderRegistration;
  private readonly providerCaps:string[];
  constructor(protected readonly store:WorkflowStore,protected readonly gate:McpDispatchGate,protected readonly host:LoopHost,provider:ProviderRegistration) {
    if(!(store.coordinator instanceof OwnerCoordinator)||!(gate instanceof McpDispatchGate)||!gate.usesStore(store)||
      [host.authenticate,host.now,host.snapshot,host.prompt,host.verification,host.policy,host.authorizeFinal].some(f=>typeof f!=="function")||!record(provider)) fail("INVALID_CONFIGURATION");
    const {invoke,...metadata}=provider, meta=copy(metadata,"INVALID_CONFIGURATION");
    if(Object.keys(meta).sort().join(",")!=="complete,id,revision,sources"||!identifier(meta.id)||!identifier(meta.revision)||typeof invoke!=="function") fail("INVALID_CONFIGURATION");
    this.provider={...meta,invoke};this.providerCaps=caps(meta.sources,meta.complete);
    this.coordinator=store.coordinator!;
    this.auth=new SecurityService({authenticate:t=>host.authenticate(t),now:()=>host.now(),resolve:()=>null});
  }
  protected check(options:Pick<LoopOptions,"deadline"|"cancelled">,expires=Number.MAX_SAFE_INTEGER):void {
    const now=this.host.now(),cancelled=options.cancelled();
    if(!integer(now)||typeof cancelled!=="boolean") fail("HOST_ERROR");
    if(cancelled) fail("CANCELLED");
    if(now>=options.deadline) fail("DEADLINE_EXCEEDED");
    if(now>=expires) fail("STALE_AUTHORITY");
  }
  private async principal(token:unknown,expected?:Principal):Promise<Principal> {
    const p=await this.auth.authenticate(token);
    if(!p.scopes.includes("models:invoke")||expected&&(p.tenantId!==expected.tenantId||p.subjectId!==expected.subjectId)) fail("FORBIDDEN");
    return p;
  }
  protected async snapshot(p:Principal,s:Record<string,unknown>):Promise<LoopAuthority> {
    const a=copy(await callback(()=>this.host.snapshot(copy(p),copy(s))),"INVALID_AUTHORITY");
    if(!record(a)||Object.keys(a).sort().join(",")!=="expires,policyVersion,providerId,providerRevision,registryRevision,releaseComplete,releaseSources,revision"||
      !identifier(a.revision)||a.policyVersion!==s.policyVersion||a.providerId!==this.provider.id||a.providerRevision!==this.provider.revision||
      a.registryRevision!==this.gate.registryRevision||!integer(a.expires)) fail("STALE_AUTHORITY");
    caps(a.releaseSources as CapabilitySource[],a.releaseComplete as boolean);
    return a as unknown as LoopAuthority;
  }
  protected async verify(p:Principal,binding:Record<string,unknown>,prompt:Envelope):Promise<void> {
    const policy=await callback(()=>this.host.verification(copy(p),copy(binding))),context=promptContext(binding);
    try {
      if(!policy||!Array.isArray(policy.keys)||policy.keys.some(k=>k.allowUnscoped!==false)) fail("PROMPT_REJECTED");
      const e=verifyEnvelope(prompt,{...policy,context,allowedAttributes:[...Object.keys(context),...this.promptAttributes()],now:this.host.now()});
      if(e.signature.sectionType!=="system"||e.signature.contentType!=="text"||![1,2].includes(e.signature.trustLevel??2)) fail("PROMPT_REJECTED");
    }catch{return fail("PROMPT_REJECTED");}
  }
  protected promptAttributes():string[] {return [];}
  protected async loadPrompt(p:Principal,binding:Record<string,unknown>,_options:LoopOptions,_reservation:OwnerReservation):Promise<Envelope> {
    const prompt=copy(await callback(()=>this.host.prompt(copy(p),copy(binding))),"PROMPT_REJECTED") as Envelope;
    await this.verify(p,binding,prompt);return prompt;
  }
  protected async inferencePrompt(_p:Principal,_binding:Record<string,unknown>,prompt:Envelope,_options:LoopOptions,_reservation:OwnerReservation):Promise<Envelope> {return prompt;}
  protected async decide(p:Principal,binding:Record<string,unknown>,data:Record<string,unknown>,phase:LoopPhase,recipientCaps:string[]):Promise<Record<string,unknown>> {
    const context={...binding,phase,dataDigest:bindingDigest(data)};
    const policy=copy(await callback(()=>this.host.policy(copy(p),copy(context),copy(data),phase)),"INVALID_POLICY");
    if(!record(policy)||Object.keys(policy).sort().join(",")!=="bindingDigest,resources"||policy.bindingDigest!==bindingDigest(context)||!Array.isArray(policy.resources)||!policy.resources.length) fail("INVALID_POLICY");
    const inputs=(policy.resources as PolicyInput[]).map(r=>{
      if(!record(r)) return fail("INVALID_POLICY");
      return {...r,capabilities:caps([{id:"recipient",capabilities:recipientCaps},{id:"path",capabilities:r.capabilities}],true)};
    });
    const decision=evaluateBatch(inputs).decision;
    if(decision==="unsupported") fail("UNSUPPORTED_POLICY");
    if(decision!=="allow") fail(phase==="release"?"OUTPUT_DENIED":"POLICY_DENIED");
    return context;
  }
  async run(token:unknown,sessionId:string,request:unknown,options:LoopOptions):Promise<Record<string,unknown>> {
    try {
      const input=copy(request);
      if(!record(input)||Object.keys(input).join(",")!=="message"||typeof input.message!=="string"||!identifier(sessionId)||
        !options||!integer(options.deadline)||typeof options.cancelled!=="function"||!integer(options.maxSteps)||options.maxSteps<1||options.maxSteps>32) fail("INVALID_REQUEST");
      options={deadline:options.deadline,cancelled:options.cancelled,maxSteps:options.maxSteps};
      const p=await this.principal(token),actor={tenantId:p.tenantId,subjectId:p.subjectId};
      this.check(options);
      const initial=await this.coordinator.runReserved(actor,async reservation=>{
        const session=await this.store.execute(actor,{action:"getSession",sessionId});
        if(session.status!=="running") fail("INACTIVE_SESSION");
        const authority=await this.snapshot(p,session);
        this.check(options,Math.min(authority.expires,session.expiresAt as number));
        const binding={tenantId:p.tenantId,subjectId:p.subjectId,sessionId,sessionVersion:session.version,nodeId:session.nodeId,nodeVersion:session.nodeVersion,
          epoch:this.store.epoch,policyVersion:authority.policyVersion,authorityRevision:authority.revision,providerId:this.provider.id,providerRevision:this.provider.revision,
          registryRevision:authority.registryRevision,deadline:options.deadline};
        const prompt=await this.loadPrompt(p,binding,options,reservation);
        return {session,authority,binding,prompt};
      });
      const {session,authority,binding}=initial,expires=Math.min(authority.expires,session.expiresAt as number);
      let prompt=initial.prompt;
      const fresh=async(checkPrompt=true)=>{
        this.check(options,expires);
        await this.principal(token,p);
        if(canonicalJson(await this.store.execute(actor,{action:"getSession",sessionId}))!==canonicalJson(session)) fail("STALE_SESSION");
        if(canonicalJson(await this.snapshot(p,session))!==canonicalJson(authority)) fail("STALE_AUTHORITY");
        if(checkPrompt)await this.verify(p,binding,prompt);
        this.check(options,expires);
      };
      const tools=await this.gate.listTools(token,sessionId,options,p);
      const messages:Record<string,unknown>[]=[{role:"system",content:prompt.data},{role:"user",content:input.message}];
      for(let step=1;step<=options.maxSteps;step++) {
        let stepBinding:Record<string,unknown>;
        const result=await this.coordinator.runReserved(actor,async reservation=>{
          await fresh(false);
          prompt=await this.inferencePrompt(p,binding,prompt,options,reservation);
          messages[0]={role:"system",content:prompt.data};
          stepBinding={...binding,step,promptDigest:bindingDigest(prompt)};
          await fresh();
          const providerRequest=copy({messages,tools});
          await this.decide(p,stepBinding,{request:providerRequest},"inference",this.providerCaps);
          await fresh();
          let raw:unknown;
          try {raw=await this.provider.invoke(copy(providerRequest),{deadline:options.deadline,cancelled:options.cancelled});}
          catch {this.check(options,expires);return fail("PROVIDER_FAILED");}
          this.check(options,expires);
          const candidate=copy(raw,"INVALID_RESPONSE");
          if(!record(candidate)) fail("INVALID_RESPONSE");
          if(candidate.type==="tool") {
            if(Object.keys(candidate).sort().join(",")!=="arguments,name,type"||typeof candidate.name!=="string"||!record(candidate.arguments)) fail("INVALID_RESPONSE");
            if(!tools.some(t=>t.name===candidate.name)) fail("TOOL_NOT_ALLOWED");
            if(step===options.maxSteps) fail("STEP_LIMIT");
            await fresh();return {candidate};
          }
          if(candidate.type!=="final"||Object.keys(candidate).sort().join(",")!=="text,type"||typeof candidate.text!=="string") fail("INVALID_RESPONSE");
          const data=copy({request:providerRequest,candidate});
          const context=await this.decide(p,stepBinding,data,"release",caps(authority.releaseSources,authority.releaseComplete));
          if(await callback(()=>this.host.authorizeFinal(copy(p),copy(context),copy(data)))!==true) fail("COMPLETION_DENIED");
          await fresh();
          return {output:{text:candidate.text,provenance:{profile:LLM_LOOP_PROFILE,providerId:this.provider.id,providerRevision:this.provider.revision,
            outputDigest:bindingDigest({text:candidate.text}),trustLevel:5,steps:step}}};
        });
        if(result.output) return result.output;
        const candidate=result.candidate!;
        let guardError:unknown;
        const output=await this.gate.callTool(token,sessionId,{name:candidate.name,arguments:candidate.arguments},{...options,
          authorizeDispatch:async(current,recipientCaps)=>{
            try {
              if(canonicalJson(current)!==canonicalJson(session)) fail("STALE_SESSION");
              await fresh();
              await this.decide(p,stepBinding!,copy({request:{messages,tools},candidate}),"tool",recipientCaps);
              await fresh();return true;
            }catch(e){guardError=e;return false;}
          }
        },p).catch(e=>{throw guardError??e;});
        // A refresh must not make a result produced under an expired prompt usable.
        await this.coordinator.run(actor,()=>fresh());
        messages.push({role:"assistant",call:{name:candidate.name,arguments:candidate.arguments}},{role:"tool",name:candidate.name,data:output.data});
        copy({messages,tools}); // Bound the accumulated transcript before its next use.
      }
      return fail("STEP_LIMIT");
    }catch(e) {
      if(e instanceof LoopError) throw e;
      if(e instanceof DispatchError) return fail(e.code);
      if(e instanceof ServiceError&&["UNAUTHENTICATED","FORBIDDEN"].includes(e.code)) return fail(e.code);
      if(e instanceof StoreError&&["NOT_FOUND","EXPIRED","STATE_BUSY"].includes(e.code)) return fail(e.code);
      return fail("HOST_ERROR");
    }
  }
}

File: implementations/python/packages/llmproxy/src/psp_cdl_llmproxy/loop.py
# SPDX-License-Identifier: Apache-2.0
"""Bounded host-embedded model/tool loop. No implicit credentials or provider SDK."""
from psp_cdl_core import canonical_json
from psp_cdl_core.crypto import verify_envelope
from psp_cdl_cdl import aggregate_capabilities, evaluate_batch
from psp_cdl_api_server.service import SecurityService, ServiceError, identifier
from psp_cdl_api_server.persistence import OwnerCoordinator, StoreError, bounded, integer
from psp_cdl_mcpproxy import McpDispatchGate, DispatchError, binding_digest

LLM_LOOP_PROFILE = "PSP-LLM-LOOP-0.1"


class LoopError(ValueError):
    def __init__(self, code):
        super().__init__(code)
        self.code = code


def copy(value, code="INVALID_REQUEST"):
    try: return bounded(value)
    except Exception: raise LoopError(code) from None


def callback(fn):
    try: return fn()
    except Exception: raise LoopError("HOST_ERROR") from None


def caps(sources, complete):
    try: return aggregate_capabilities(sources, complete)
    except Exception: raise LoopError("INVALID_CAPABILITIES") from None


def prompt_context(binding):
    """Required signed attributes; never include the host binding in model messages."""
    fields = {"tenant-id":"tenantId", "subject-id":"subjectId", "session-id":"sessionId", "session-version":"sessionVersion",
              "node-id":"nodeId", "node-version":"nodeVersion", "store-epoch":"epoch", "policy-version":"policyVersion",
              "authority-revision":"authorityRevision", "provider-id":"providerId", "provider-revision":"providerRevision", "registry-revision":"registryRevision"}
    return {attribute:str(binding[field]) for attribute,field in fields.items()}


class BufferedLlmLoop:
    def __init__(self, store, gate, host, provider):
        if not isinstance(store.coordinator, OwnerCoordinator) or not isinstance(gate, McpDispatchGate) or not gate.uses_store(store) or any(
            not callable(getattr(host, name, None)) for name in ("authenticate", "now", "snapshot", "prompt", "verification", "policy", "authorize_final")
        ) or type(provider) is not dict:
            raise LoopError("INVALID_CONFIGURATION")
        invoke = provider.get("invoke")
        meta = copy({k:v for k,v in provider.items() if k != "invoke"}, "INVALID_CONFIGURATION")
        if set(meta) != {"id", "revision", "sources", "complete"} or not identifier(meta["id"]) or not identifier(meta["revision"]) or not callable(invoke):
            raise LoopError("INVALID_CONFIGURATION")
        self._provider = {**meta, "invoke":invoke}
        self._provider_caps = caps(meta["sources"], meta["complete"])
        self._store, self._gate, self._host = store, gate, host
        self._coordinator, self._auth = store.coordinator, SecurityService(host)

    def _check(self, options, expires=9007199254740991):
        now, cancelled = self._host.now(), options["cancelled"]()
        if not integer(now) or type(cancelled) is not bool: raise LoopError("HOST_ERROR")
        if cancelled: raise LoopError("CANCELLED")
        if now >= options["deadline"]: raise LoopError("DEADLINE_EXCEEDED")
        if now >= expires: raise LoopError("STALE_AUTHORITY")

    def _principal(self, token, expected=None):
        p = self._auth.authenticate(token)
        if "models:invoke" not in p["scopes"] or expected and any(p[k] != expected[k] for k in ("tenantId", "subjectId")):
            raise LoopError("FORBIDDEN")
        return p

    def _snapshot(self, p, session):
        a = copy(callback(lambda:self._host.snapshot(copy(p), copy(session))), "INVALID_AUTHORITY")
        if type(a) is not dict or set(a) != {"revision", "policyVersion", "providerId", "providerRevision", "registryRevision", "expires", "releaseSources", "releaseComplete"} or not identifier(a["revision"]) or a["policyVersion"] != session["policyVersion"] or a["providerId"] != self._provider["id"] or a["providerRevision"] != self._provider["revision"] or a["registryRevision"] != self._gate.registry_revision or not integer(a["expires"]):
            raise LoopError("STALE_AUTHORITY")
        caps(a["releaseSources"], a["releaseComplete"])
        return a

    def _verify(self, p, binding, prompt):
        policy = callback(lambda:self._host.verification(copy(p), copy(binding)))
        context = prompt_context(binding)
        try:
            if type(policy) is not dict or type(policy.get("keys")) is not list or any(k.get("allowUnscoped") is not False for k in policy["keys"]):
                raise LoopError("PROMPT_REJECTED")
            e = verify_envelope(prompt, {**policy, "context":context, "allowedAttributes":list(context), "now":self._host.now()})
            if e["signature"]["sectionType"] != "system" or e["signature"]["contentType"] != "text" or e["signature"].get("trustLevel", 2) not in (1, 2):
                raise LoopError("PROMPT_REJECTED")
        except Exception: raise LoopError("PROMPT_REJECTED") from None

    def _decide(self, p, binding, data, phase, recipient_caps):
        context = {**binding, "phase":phase, "dataDigest":binding_digest(data)}
        policy = copy(callback(lambda:self._host.policy(copy(p), copy(context), copy(data), phase)), "INVALID_POLICY")
        if type(policy) is not dict or set(policy) != {"bindingDigest", "resources"} or policy["bindingDigest"] != binding_digest(context) or type(policy["resources"]) is not list or not policy["resources"]:
            raise LoopError("INVALID_POLICY")
        inputs = []
        for resource in policy["resources"]:
            if type(resource) is not dict: raise LoopError("INVALID_POLICY")
            inputs.append({**resource, "capabilities":caps([{"id":"recipient", "capabilities":recipient_caps}, {"id":"path", "capabilities":resource.get("capabilities")}], True)})
        decision = evaluate_batch(inputs)["decision"]
        if decision == "unsupported": raise LoopError("UNSUPPORTED_POLICY")
        if decision != "allow": raise LoopError("OUTPUT_DENIED" if phase == "release" else "POLICY_DENIED")
        return context

    def run(self, token, session_id, request, options):
        try:
            value = copy(request)
            if type(value) is not dict or set(value) != {"message"} or type(value["message"]) is not str or not identifier(session_id) or type(options) is not dict or not integer(options.get("deadline")) or not callable(options.get("cancelled")) or not integer(options.get("maxSteps")) or not 1 <= options["maxSteps"] <= 32:
                raise LoopError("INVALID_REQUEST")
            options = {k:options[k] for k in ("deadline", "cancelled", "maxSteps")}
            p = self._principal(token)
            actor = {k:p[k] for k in ("tenantId", "subjectId")}
            self._check(options)
            def initialize():
                session = self._store.execute(actor, {"action":"getSession", "sessionId":session_id})
                if session["status"] != "running": raise LoopError("INACTIVE_SESSION")
                authority = self._snapshot(p, session)
                self._check(options, min(authority["expires"], session["expiresAt"]))
                binding = {"tenantId":p["tenantId"], "subjectId":p["subjectId"], "sessionId":session_id, "sessionVersion":session["version"],
                           "nodeId":session["nodeId"], "nodeVersion":session["nodeVersion"], "epoch":self._store.epoch, "policyVersion":authority["policyVersion"],
                           "authorityRevision":authority["revision"], "providerId":self._provider["id"], "providerRevision":self._provider["revision"],
                           "registryRevision":authority["registryRevision"], "deadline":options["deadline"]}
                prompt = copy(callback(lambda:self._host.prompt(copy(p), copy(binding))), "PROMPT_REJECTED")
                self._verify(p, binding, prompt)
                return session, authority, binding, prompt
            session, authority, binding, prompt = self._coordinator.run(actor, initialize)
            expires = min(authority["expires"], session["expiresAt"])
            def fresh():
                self._check(options, expires)
                self._principal(token, p)
                if canonical_json(self._store.execute(actor, {"action":"getSession", "sessionId":session_id})) != canonical_json(session): raise LoopError("STALE_SESSION")
                if canonical_json(self._snapshot(p, session)) != canonical_json(authority): raise LoopError("STALE_AUTHORITY")
                self._verify(p, binding, prompt)
                self._check(options, expires)
            tools = self._gate.list_tools(token, session_id, options, p)
            messages = [{"role":"system", "content":prompt["data"]}, {"role":"user", "content":value["message"]}]
            for step in range(1, options["maxSteps"] + 1):
                step_binding = {**binding, "step":step, "promptDigest":binding_digest(prompt)}
                def infer():
                    fresh()
                    provider_request = copy({"messages":messages, "tools":tools})
                    self._decide(p, step_binding, {"request":provider_request}, "inference", self._provider_caps)
                    fresh()
                    try: raw = self._provider["invoke"](copy(provider_request), {"deadline":options["deadline"], "cancelled":options["cancelled"]})
                    except Exception:
                        self._check(options, expires)
                        raise LoopError("PROVIDER_FAILED") from None
                    self._check(options, expires)
                    candidate = copy(raw, "INVALID_RESPONSE")
                    if type(candidate) is not dict: raise LoopError("INVALID_RESPONSE")
                    if candidate.get("type") == "tool":
                        if set(candidate) != {"type", "name", "arguments"} or type(candidate["name"]) is not str or type(candidate["arguments"]) is not dict: raise LoopError("INVALID_RESPONSE")
                        if not any(t["name"] == candidate["name"] for t in tools): raise LoopError("TOOL_NOT_ALLOWED")
                        if step == options["maxSteps"]: raise LoopError("STEP_LIMIT")
                        fresh()
                        return {"candidate":candidate}
                    if candidate.get("type") != "final" or set(candidate) != {"type", "text"} or type(candidate["text"]) is not str: raise LoopError("INVALID_RESPONSE")
                    data = copy({"request":provider_request, "candidate":candidate})
                    context = self._decide(p, step_binding, data, "release", caps(authority["releaseSources"], authority["releaseComplete"]))
                    if callback(lambda:self._host.authorize_final(copy(p), copy(context), copy(data))) is not True: raise LoopError("COMPLETION_DENIED")
                    fresh()
                    return {"output":{"text":candidate["text"], "provenance":{"profile":LLM_LOOP_PROFILE, "providerId":self._provider["id"],
                            "providerRevision":self._provider["revision"], "outputDigest":binding_digest({"text":candidate["text"]}), "trustLevel":5, "steps":step}}}
                result = self._coordinator.run(actor, infer)
                if "output" in result: return result["output"]
                candidate = result["candidate"]
                guard_error = []
                def authorize_dispatch(current, recipient_caps):
                    try:
                        if canonical_json(current) != canonical_json(session): raise LoopError("STALE_SESSION")
                        fresh()
                        self._decide(p, step_binding, copy({"request":{"messages":messages, "tools":tools}, "candidate":candidate}), "tool", recipient_caps)
                        fresh()
                        return True
                    except Exception as exc:
                        guard_error.append(exc)
                        return False
                try:
                    output = self._gate.call_tool(token, session_id, {"name":candidate["name"], "arguments":candidate["arguments"]}, {**options, "authorizeDispatch":authorize_dispatch}, p)
                except Exception:
                    if guard_error: raise guard_error[0]
                    raise
                messages.extend([{"role":"assistant", "call":{"name":candidate["name"], "arguments":candidate["arguments"]}}, {"role":"tool", "name":candidate["name"], "data":output["data"]}])
                copy({"messages":messages, "tools":tools})
            raise LoopError("STEP_LIMIT")
        except LoopError: raise
        except DispatchError as exc: raise LoopError(exc.code) from None
        except ServiceError as exc: raise LoopError(exc.code if exc.code in ("UNAUTHENTICATED", "FORBIDDEN") else "HOST_ERROR") from None
        except StoreError as exc: raise LoopError(exc.code if exc.code in ("NOT_FOUND", "EXPIRED", "STATE_BUSY") else "HOST_ERROR") from None
        except Exception: raise LoopError("HOST_ERROR") from None
